src/runtime/value.cpp

```diff
@@ -65,50 +66,39 @@ std::string HeapObject::to_str() const {
   return str.str();
 }
 
-String::String(size_t length_) : length(length_) {}
+String::String(size_t length_) : length(length_) {
+  char *str = static_cast<char *>(data());
+  std::fill(str, str + length + 1, 0);
+}
 
 String::String(const String &s) : length(s.length) { memcpy(data(), s.data(), length + 1); }
+String::String(const char *str, size_t length) : length(length) { memcpy(data(), str, length + 1); }
 
 String *String::claim(Heap &h, size_t length) {
   return new (h.claim(reserve(length))) String(length);
 }
 
 String *String::claim(Heap &h, const std::string &str) {
-  String *out = claim(h, str.size());
-  memcpy(out->c_str(), str.c_str(), str.size() + 1);
-  return out;
+  return new (h.claim(reserve(str.length()))) String(str.c_str(), str.length());
 }
 
 String *String::claim(Heap &h, const char *str, size_t length) {
-  auto out = claim(h, length);
-  memcpy(out->c_str(), str, length);
-  out->c_str()[length] = 0;
-  return out;
+  return new (h.claim(reserve(length))) String(str, length);
 }
 
 String *String::alloc(Heap &h, size_t length) {
   return new (h.alloc(reserve(length))) String(length);
 }
 
 String *String::alloc(Heap &h, const std::string &str) {
-  String *out = alloc(h, str.size());
-  memcpy(out->data(), str.c_str(), str.size() + 1);
-  return out;
+  return new (h.alloc(reserve(str.length()))) String(str.c_str(), str.length());
 }
 
 String *String::alloc(Heap &h, const char *str, size_t length) {
-  String *out = alloc(h, length);
-  memcpy(out->c_str(), str, length);
-  out->c_str()[length] = 0;
-  return out;
+  return new (h.alloc(reserve(length))) String(str, length);
 }
```

Comment:
I wonder if this was being used somewhere where the `out->c_str()[length] = 0;` was required because `str` didn't have a null term in some use case?

Reply:
omg....what a roller coaster. This upsets me.